auth0pull/auth0pull.go

```diff
@@ -0,0 +1,183 @@
+package auth0pull
+
+import (
+	"context"
+	"encoding/json"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/mozilla-services/foxsec-pipeline-contrib/common"
+
+	"cloud.google.com/go/datastore"
+	stackdriver "cloud.google.com/go/logging"
+	log "github.com/sirupsen/logrus"
+	"go.mozilla.org/mozlogrus"
+	"gopkg.in/auth0.v1/management"
+)
+
+const (
+	LASTLOGID_KIND      = "last_log_id_auth0"
+	LASTLOGID_KEY       = "last_log_id_auth0"
+	LASTLOGID_NAMESPACE = "last_log_id_auth0"
+
+	LOGGER_NAME = "auth0pull"
+)
+
+var (
+	PROJECT_ID string
+
+	config = &common.Configuration{}
+
+	logClient         *management.LogManager
+	datastoreClient   *datastore.Client
+	stackdriverClient *stackdriver.Client
+)
+
+func init() {
+	mozlogrus.Enable("auth0pull")
+	PROJECT_ID = os.Getenv("GCP_PROJECT")
+	InitConfig()
+}
+
+func InitConfig() {
+	log.Info("Starting up...")
+	configPath := os.Getenv("CONFIG_PATH")
+	if configPath == "" {
+		log.Fatal("$CONFIG_PATH must be set.")
+	}
+	err := config.LoadFrom(configPath)
+	if err != nil {
+		log.Fatalf("Could not load config file from `%s`: %s", configPath, err)
+	}
+
+	m, err := management.New(config.Auth0Domain, config.Auth0ClientId, config.Auth0ClientSecret)
+	if err != nil {
+		log.Fatalf("Could not create management client: %s", err)
+	}
+	logClient = management.NewLogManager(m)
+
+	stackdriverClient, err = stackdriver.NewClient(context.Background(), PROJECT_ID)
+	if err != nil {
+		log.Fatalf("Could not create stackdriver client: %s", err)
+	}
+
+	datastoreClient, err = datastore.NewClient(context.Background(), PROJECT_ID)
+	if err != nil {
+		log.Fatalf("Could not create datastore client: %s", err)
+	}
+}
+
+func getLogs(from string) ([]*management.Log, error) {
+	var collectedLogs []*management.Log
+
+	for {
+		logs, err := logClient.List(func(v url.Values) {
+			v.Set("from", from)
+			v.Set("take", "100")
```

Comment:
Is this the number of grab (just assuming)? Should this be higher if so, is it possible to fall behind here?

Reply:
>Is this the number of grab (just assuming)?

Yes

>Should this be higher if so

100 is the max (https://auth0.com/docs/logs#limitations)

>is it possible to fall behind here?

I don't think so, but (I think) this will need to be tested in production to be sure. The idea with this loop is to grab all logs (100 at a time) up to the latest, then save the latest as the next place to start from. If it's running every 5 minutes, as long as it can get all logs from the last 5 minutes in under 5 minutes, we should be good.

Reply:
👌